# Importy:

In [91]:
import numpy as np
from scipy.linalg import solve, inv, det

___
### Funkcja eliminacji Gaussa
Zaimpelementowana przeze mnie funkcja `gauss(matrix)` pobiera daną macierz, następnie ustala liczbę kolumn, i  w każdej z nich szuka elementu podstawowego, czyli maksymalnego w danej kolumnie. Jeśli jest to konieczne, zamienia wiersze miejscami. Następnie, zgodnie z algorytmem eliminacji Gaussa, eliminuje elementy poniżej głównej przekątnej, zwracając macierz górnotrójkątną. 


In [77]:
def gauss(matrix):
    n = matrix.shape[0]
    for i in range(n):
        max_row = np.argmax(np.abs(matrix[i:, i])) + i

        if max_row != i:
            matrix[[i, max_row]] = matrix[[max_row, i]]

        for j in range(i + 1, n):
            ratio = matrix[j, i] / matrix[i, i]
            matrix[j] -= ratio * matrix[i]
    
    return matrix

___
### Funkcja rozwiązywania układów równań w postaci macierzowej

Zaimplementowana funkcja `solution(A, b)` na początku łączy macierz współczynników $A$ z wektorem wyników $\vec{b}$, tworząc jedną macierz rozszerzoną. Następnie wywołuje funkcję `gauss(matrix)`, aby rozwiązać układ równań metodą eliminacji Gaussa. Następnie, przy użyciu metody wstecznej eliminacji, iteruje przez wiersze od dołu do góry, obliczając wartość każdej zmiennej. Funkcja końcowo zwraca wektor rozwiązań $\vec{x}$.

In [56]:
def solution(A, b):
    matrix = np.hstack([A, b.reshape(-1, 1)])
    gauss(matrix)
    n = matrix.shape[0]
    x = np.zeros(n)
    
    for i in range(n - 1, -1, -1):
        x[i] = (matrix[i, -1] - np.dot(matrix[i, i + 1:n], x[i + 1:n])) / matrix[i, i]
    
    return x

___
### Funkcja licząca wyznacznik macierzy
Zaimplementowana funkcja `determinant(matrix)` oblicza wyznacznik macierzy kwadratowej za pomocą metody eliminacji Gaussa. 
Funkcja przeprowadza eliminację Gaussa, zliczając przy tym zamiany wierszy. Jeśli napotka element diagonalny równy zero, zwraca wyznacznik równy zeru.
Wyznacznik obliczany jest jako iloczyn elementów na głównej przekątnej macierzy. Wynik zwracany jest z uwzględnieniem, że jeśli liczba zamian wierszy jest nieparzysta, znak wyznacznika jest zmieniany na przeciwny.

In [ ]:
def determinant(matrix):
    n = matrix.shape[0]
    det = 1
    row_swaps = 0  
    
    for i in range(n):
        max_row = np.argmax(np.abs(matrix[i:, i])) + i

        if max_row != i:
            matrix[[i, max_row]] = matrix[[max_row, i]]
            row_swaps += 1  
        
        if matrix[i, i] == 0:
            return 0

        for j in range(i + 1, n):
            ratio = matrix[j, i] / matrix[i, i]
            matrix[j] -= ratio * matrix[i]

    for i in range(n):
        det *= matrix[i, i]

    if row_swaps % 2 == 1:
        det *= -1  
    
    return det

___
### Funkcja mnożąca dwie macierze
Zaimplementowana przeze mnie funkcja `multiply(A, B)` wykonuje mnożenie dwóch macierzy: $A$ oraz $B$. Na początku funkcja ustala liczby wierszy i kolumn macierzy oraz sprawdza, czy liczba kolumn w macierzy $A$ jest zgodna z liczbą wierszy w macierzy $B$.
Następnie tworzy macierz wynikową o wymiarach odpowiadających liczbie wierszy z $A$ i liczbie kolumn z $B$. 
Funkcja wykonuje mnożenie macierzy poprzez zagnieżdżoną pętlę: dla każdego wiersza macierzy $A$ i każdej kolumny macierzy $B$, sumuje iloczyny odpowiednich elementów, aby uzyskać elementy macierzy wynikowej, która jest na końcu zwracana.

In [92]:
def multiply(A, B):
    rows_A = len(A)
    cols_A = len(A[0])
    rows_B = len(B)
    if cols_A != rows_B:
        raise ValueError('Niezgodne wymiary')
    cols_B = len(B[0])

    result = [[0 for _ in range(cols_B)] for _ in range(rows_A)]

    for i in range(rows_A):
        for j in range(cols_B):
            for k in range(cols_A):
                result[i][j] += A[i][k] * B[k][j]

    return result

___
### Funkcja obliczająca macierz odwrotną


Funkcja `inverse(matrix)` oblicza macierz odwrotną dla danej macierzy kwadratowej. Na początku funkcja sprawdza, czy wyznacznik macierzy jest równy zero. Jeśli tak, funkcja zgłosi błąd, ponieważ nie istnieje macierz odwrotna do macierzy osobliwej.
Następnie tworzy macierz rozszerzoną o macierz jednostkową. W dalszej kolejności funkcja przeprowadza algorytm eliminacji Gaussa, lecz w tym przypadku, chcemy aby macierz A została zamieniona na macierz jednostkową. 
Na koniec funkcja zwraca prawą część macierzy rozszerzonej, która zawiera obliczoną macierz odwrotną.

In [88]:
def inverse(matrix):
    if determinant(matrix) == 0:
        raise ValueError('Macierz osobliwa, nie ma macierzy odwrotnej')
    n = A.shape[0]
    matrix = np.hstack([matrix, np.eye(n)])
    
    for i in range(n):
        max_row = np.argmax(np.abs(matrix[i:, i])) + i
        
        if max_row != i:
            matrix[[i, max_row]] = matrix[[max_row, i]]
        
        matrix[i] /= matrix[i, i]

        for j in range(n):
            if j != i:
                matrix[j] -= matrix[i] * matrix[j, i]

    A_inv = matrix[:, n:]
    
    return A_inv


___
## Zadanie 4

In [57]:
A = np.array([[0, 0, 2, 1, 2],
              [0, 1, 0, 2, -1],
              [1, 2, 0, -2, 0],
              [0, 0, 0, -1, 1],
              [0, 1, -1, 1, -1]], dtype=float)

b = np.array([1, 1, -4, -2, -1], dtype=float)


print('Wektor rozwiązań obliczony metodą eliminacji Gaussa to:',solution(A,b))

print('\nWektor rozwiązań uzyskany przez bibliotekę scipy to:', solve(A,b))

Wektor rozwiązań obliczony metodą eliminacji Gaussa to: [ 2. -2.  1.  1. -1.]

Wektor rozwiązań uzyskany przez bibliotekę scipy to: [ 2. -2.  1.  1. -1.]


___
## Zadanie 5

Współczynniki wielomianu $ y = a_0 + a_1 x + a_2 x^2 + a_3 x^3 + a_4 x^4$, który przechodzi przez punkty $(0,1) , (1,1) , (3,3) , (5,2) , (6,-2) $ wyznaczono, poprzez rozwiązanie układu równań: 
$$
\begin{align*}
\left\{
\begin{aligned}
a_0 &= -1 \\
a_0 + a_1 + a_2 + a_3 + a_4 &= 1 \\
a_0 + 3a_1 + 9a_2 + 27a_3 + 81a_4 &= 3 \\
a_0 + 5a_1 + 25a_2 + 125a_3 + 625a_4 &= 2 \\
a_0 + 6a_1 + 36a_2 + 216a_3 + 1296a_4 &= -2 \\
\end{aligned}
\right.
\end{align*}

$$
W wersji macierzowej ten układ równań to:
$$
\begin{align*}
\begin{bmatrix}
1 & 0 & 0 & 0 & 0 \\
1 & 1 & 1 & 1 & 1 \\
1 & 3 & 9 & 27 & 81 \\
1 & 5 & 25 & 125 & 625 \\
1 & 6 & 36 & 216 & 1296 \\
\end{bmatrix}
\begin{bmatrix}
a_0 \\
a_1 \\
a_2 \\
a_3 \\
a_4 \\
\end{bmatrix}
=
\begin{bmatrix}
-1 \\
1 \\
3 \\
2 \\
-2 \\
\end{bmatrix}
\end{align*}

$$

In [58]:
A = np.array([
    [1, 0, 0, 0, 0],
    [1, 1, 1, 1, 1],
    [1, 3, 9, 27, 81],
    [1, 5, 25, 125, 625],
    [1, 6, 36, 216, 1296]], dtype=float)

b = np.array([-1, 1, 3, 2, -2], dtype=float)

print('Wektor rozwiązań obliczony metodą eliminacji Gaussa to:',solution(A,b))

print('\nWektor rozwiązań uzyskany przez bibliotekę scipy to:', solve(A,b))

Wektor rozwiązań obliczony metodą eliminacji Gaussa to: [-1.          2.68333333 -0.875       0.21666667 -0.025     ]

Wektor rozwiązań uzyskany przez bibliotekę scipy to: [-1.          2.68333333 -0.875       0.21666667 -0.025     ]


Z tego wynika, że współczynniki wielomianu to:
$
a_0 = -1 ,
a_1 \approx 2.683,
a_2 = -0.875,
a_3 \approx 0.217,
a_4 = -0.025,
$

___
## Zadanie 6


In [75]:
A = np.array([[3.50, 2.77, -0.76, 1.80],
              [-1.80, 2.68, 3.44, -0.09],
              [0.27, 5.07, 6.90, 1.61],
              [1.71, 5.45, 2.68, 1.71]], dtype=float)

b = np.array([7.31, 4.23, 13.85, 11.55], dtype=float)

print('Wektor rozwiązań obliczony metodą eliminacji Gaussa to:',solution(A,b))

print('\nWektor rozwiązań uzyskany przez bibliotekę scipy to:', solve(A,b))

print('\nWyznacznik obliczony metodą eliminacji Gaussa to:',determinant(A))

print('\nWyznacznik uzyskany przez bibliotekę scipy to:', det(A))

x_vector = [[value] for value in solution(A, b)]  

print('\nWynik mnożenia obliczony za pomocą samodzielnej implementacji:', multiply(A, x_vector))
print('\nWynik mnożenia uzyskany za pomocą biblioteki numpy:\n', np.dot(A,x_vector))


Wektor rozwiązań obliczony metodą eliminacji Gaussa to: [1. 1. 1. 1.]

Wektor rozwiązań uzyskany przez bibliotekę scipy to: [1. 1. 1. 1.]

Wyznacznik obliczony metodą eliminacji Gaussa to: -0.22579734000000884

Wyznacznik uzyskany przez bibliotekę scipy to: -0.22579733999999008

Wynik mnożenia obliczony za pomocą samodzielnej implementacji: [[7.31], [4.23], [13.849999999999998], [11.55]]

Wynik mnożenia uzyskany za pomocą biblioteki numpy:
 [[ 7.31]
 [ 4.23]
 [13.85]
 [11.55]]


Wektor rozwiązań równania to: $ \vec{x} =  [1,1,1,1]$. 

Wyznacznik macierzy to: $\text{det}A \approx -0.226$

$A \vec{x} = \begin{bmatrix}
7.31 \\
4.23 \\
13.85 \\
11.55 \\
\end{bmatrix} $

Uzyskane rozwiązania nie są całkowicie dokładne. Jest to spowodowane niedokładnościa w interpretacji liczb przez komputer.

___
## Zadanie 7

In [63]:
A = np.array([[10, -2, -1, 2, 3, 1, -4, 7],
              [5, 11, 3, 10, -3, 3, 3, -4],
              [7, 12, 1, 5, 3, -12, 2, 3],
              [8, 7, -2, 1, 3, 2, 2, 4],
              [2, -15, -1, 1, 4, -1, 8, 3],
              [4, 2, 9, 1, 12, -1, 4, 1],
              [-1, 4, -7, -1, 1, 1, -1, -3],
              [-1, 3, 4, 1, 3, -4, 7, 6]], dtype=float)

b = np.array([0,12,-5,3,-25,-26,9,-7], dtype=float)

print('Wektor rozwiązań obliczony metodą eliminacji Gaussa to:',solution(A,b))

print('\nWektor rozwiązań uzyskany przez bibliotekę scipy to:', solve(A,b))


Wektor rozwiązań obliczony metodą eliminacji Gaussa to: [-1.  1. -1.  1. -1.  1. -1.  1.]

Wektor rozwiązań uzyskany przez bibliotekę scipy to: [-1.  1. -1.  1. -1.  1. -1.  1.]


$\vec{x} = [-1, 1, -1, 1, -1, 1, -1, 1]$

___
## Zadanie 8

In [89]:
A = np.array([[ 2, -1,  0,  0,  0,  0],
              [-1,  2, -1,  0,  0,  0],
              [ 0, -1,  2, -1,  0,  0],
              [ 0,  0, -1,  2, -1,  0],
              [ 0,  0,  0, -1,  2, -1],
              [ 0,  0,  0,  0, -1,  5]], dtype=float)


A_inv = inverse(A)
print('Macierz odwrotna obliczona przez zaimplementowaną metodę Gaussa:')
print(A_inv)

A_inv_scipy = inv(A)
print('\nMacierz odwrotna uzyskana za pomocą biblioteki scipy:')
print(A_inv_scipy)



Macierz odwrotna obliczona przez zaimplementowaną metodę Gaussa:
[[0.84 0.68 0.52 0.36 0.2  0.04]
 [0.68 1.36 1.04 0.72 0.4  0.08]
 [0.52 1.04 1.56 1.08 0.6  0.12]
 [0.36 0.72 1.08 1.44 0.8  0.16]
 [0.2  0.4  0.6  0.8  1.   0.2 ]
 [0.04 0.08 0.12 0.16 0.2  0.24]]

Macierz odwrotna uzyskana za pomocą biblioteki scipy:
[[0.84 0.68 0.52 0.36 0.2  0.04]
 [0.68 1.36 1.04 0.72 0.4  0.08]
 [0.52 1.04 1.56 1.08 0.6  0.12]
 [0.36 0.72 1.08 1.44 0.8  0.16]
 [0.2  0.4  0.6  0.8  1.   0.2 ]
 [0.04 0.08 0.12 0.16 0.2  0.24]]


Macierz $A^{-1}$ jest symetryczna, lecz nie jest trójdiagonalna. Oznacza to, że macierz odwrotna do macierzy trójdiagonalnej nie musi być macierzą trójdiagonalną. 

___
## Zadanie 9

In [90]:
A = np.array([[1, 3, -9, 6, 4],
              [2, -1, 6, 7, 1],
              [3, 2, -3, 15, 5],
              [8, -1, 1, 4, 2],
              [11, 1, -2, 18, 7]], dtype=float)


A_inv = inverse(A)
print('Macierz odwrotna obliczona przez zaimplementowaną metodę Gaussa:')
print(A_inv)

A_inv_scipy = inv(A)
print('\nMacierz odwrotna uzyskana za pomocą biblioteki scipy:')
print(A_inv_scipy)



Macierz odwrotna obliczona przez zaimplementowaną metodę Gaussa:
[[-2.22649869e+15 -2.22649869e+15 -2.22649869e+15 -4.45299738e+15
   4.45299738e+15]
 [-1.05758688e+16 -1.05758688e+16 -1.05758688e+16 -2.11517376e+16
   2.11517376e+16]
 [-1.77107851e+15 -1.77107851e+15 -1.77107851e+15 -3.54215701e+15
   3.54215701e+15]
 [ 0.00000000e+00  0.00000000e+00  1.00000000e+00  1.00000000e+00
  -1.00000000e+00]
 [ 4.50359963e+15  4.50359963e+15  4.50359963e+15  9.00719925e+15
  -9.00719925e+15]]

Macierz odwrotna uzyskana za pomocą biblioteki scipy:
[[-5.56624673e+14 -5.56624673e+14 -5.56624673e+14 -1.11324935e+15
   1.11324935e+15]
 [-2.64396720e+15 -2.64396720e+15 -2.64396720e+15 -5.28793439e+15
   5.28793439e+15]
 [-4.42769626e+14 -4.42769626e+14 -4.42769626e+14 -8.85539253e+14
   8.85539253e+14]
 [-2.50000000e-01 -2.50000000e-01  7.50000000e-01  5.00000000e-01
  -5.00000000e-01]
 [ 1.12589991e+15  1.12589991e+15  1.12589991e+15  2.25179981e+15
  -2.25179981e+15]]


Rozwiązanie w tym zadaniu nie jest poprawne, ponieważ niedokładność w zapisywaniu wartości przez komputer, powoduje, że program nie uzyskuje dokładnego  wyniku wyznacznika, czyli 0, lecz liczbę bardzo bliską tej wartości. Powoduje to błędne wyliczenie macierzy odwrotnej, która dla tego konkretnego przypadku nie istnieje, ponieważ macierz A jest osobliwa.